# Day 2 String Cleaning - from VCI Complete Roadmap

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Load the data
df = pd.read_csv('nigeria_messy_sales_dataset.csv')

In [3]:
# ALWAYS start with inspection
print(f"Dataset shape: {df.shape}")
print(f"\nColumn names and types:")
print(df.dtypes)
print("\nFirst few rows:")
print(df.head())
print("\nInfo:")
print(df.info())

Dataset shape: (550, 9)

Column names and types:
Customer Name     object
State             object
Product           object
Units Sold       float64
Unit Price       float64
Total Sale       float64
Sale Date         object
Sales Channel     object
Order ID          object
dtype: object

First few rows:
     Customer Name    State     Product  Units Sold  Unit Price  Total Sale  \
0              NaN   rivers    KEYBOARD         NaN         NaN         NaN   
1     Allison Hill    Lagos  Headphones         NaN   267992.94         NaN   
2      Noah Rhodes  Anambra    Keyboard         NaN    42364.41         NaN   
3  Angie Henderson    Delta    Keyboard         NaN   279444.94         NaN   
4    Daniel Wagner    Delta      Tablet         NaN    95899.74         NaN   

    Sale Date Sales Channel                              Order ID  
0  2025-04-27        Online                                   NaN  
1  2024-03-15     Wholesale  4c636e95-025f-4543-8997-623ae0723d96  
2  2024-12-10   

# TASK 1: Standardize State Names

## Step 1: Explore the messiness

In [12]:
print(f"Unique states BEFORE cleaning:\n{df['State'].unique()}")
print(f"\nNumber of unique state names BEFORE cleaining:\n{len(df['State'].unique())}")

Unique states BEFORE cleaning:
['rivers' 'Lagos' 'Anambra' 'Delta' 'Oyo' 'Rivers' 'Imo' 'Kano' 'plateau'
 'Niger' 'Abuja' 'Enugu' 'Bauchi' 'Katsina' 'lagos' 'Ekiti' 'Plateau'
 'Kaduna' 'Sokoto' 'kano' 'Benue' 'Osun' 'imo' 'Borno' 'enugu' 'anambra'
 'katsina' 'delta' 'Cross River' 'oyo' 'osun' 'kaduna' 'niger' 'borno'
 'abuja' 'bauchi' 'cross river' 'ekiti' 'sokoto' 'benue']

Number of unique state names BEFORE cleaining:
40


## Step 2: Create standardization plan

Everything should be in title case due to the data being state names (proper nouns).

## Step 3: Implement cleaning 

In [15]:
# Clean the data, make all state names
df['State'] = df['State'].str.title()

# Verify the data
print(f"Unique state names AFTER cleaning:\n{df['State'].unique()}")
print(f"\nNumber of unique state names AFTER cleaning:\n{len(df['State'].unique())}")

Unique state names AFTER cleaning:
['Rivers' 'Lagos' 'Anambra' 'Delta' 'Oyo' 'Imo' 'Kano' 'Plateau' 'Niger'
 'Abuja' 'Enugu' 'Bauchi' 'Katsina' 'Ekiti' 'Kaduna' 'Sokoto' 'Benue'
 'Osun' 'Borno' 'Cross River']

Number of unique state names AFTER cleaning:
20


# TASK 2: Clean Product Names

## Step 1: Explore

In [16]:
print(f"Unique product names BEFORE cleaning:\n{df['Product'].unique()}")
print(f"\nNumber of unique product names BEFORE cleaning:\n{len(df['Product'].unique())}")

Unique product names BEFORE cleaning:
['KEYBOARD' 'Headphones' 'Keyboard' 'Tablet' 'Camera' 'CAMERA' 'Laptop'
 'Charger' 'CHARGER' 'Monitor' 'LAPTOP' 'Phone' 'MONITOR' 'PHONE' 'TABLET'
 'HEADPHONES']

Number of unique product names BEFORE cleaning:
16


## Step 2: Create standardization plan
I'm thinking just change everything to title case. There's no need for all capitalization and titlecase just reads nicer.

## Step 3: Implement cleaning

In [18]:
# Clean the product data
df['Product'] = df['Product'].str.title()

# Verify cleaning
print(f"Unique product names AFTER cleaning:\n{df['Product'].unique()}")
print(f"\nNumber of unique product names after cleaning:\n{len(df['Product'].unique())}")

Unique product names AFTER cleaning:
['Keyboard' 'Headphones' 'Tablet' 'Camera' 'Laptop' 'Charger' 'Monitor'
 'Phone']

Number of unique product names after cleaning:
8


# TASK 3: Standardize Sales Channels

## Task 1: Explore data

In [38]:
print(f"Sales channels values BEFORE:\n{df['Sales Channel'].unique()}")
print(f"\nUnique count:\n{df['Sales Channel'].nunique()}")
print(f"\nValue counts:\n{df['Sales Channel'].value_counts()}")
print(f"\nNull value count:\n{df['Sales Channel'].isnull().sum()} ({(df['Sales Channel'].isnull().sum()/len(df) * 100):.2f})%")

Sales channels values BEFORE:
['Online' 'Wholesale' nan 'Direct' 'Retail']

Unique count:
4

Value counts:
Sales Channel
Direct       124
Wholesale    111
Retail       108
Online       101
Name: count, dtype: int64

Null value count:
106 (19.27)%


## Step 2: Think about it
Dropping 19% of data is a lot, and probably a bad business decision. So for now we will leave the values as NaN. Possibly later down the road we can do a weighted analysis of which orders come from which channels and add assumed values then, but not now.

## Step 3: Implement cleaning
We are leaving the null data as NaN for now, and the rest of the data is already standardized in titlecase, so I would say our work here is done.

# TASK 4: Handle Missing Customer Names 

## Step 1: Explore the problem

In [45]:
print("Missing customer names:")
print(f"Missing count: {df['Customer Name'].isnull().sum()} ({df['Customer Name'].isnull().sum() / len(df) *100:.2f})%")


Missing customer names:
Missing count: 43 (7.82)%


## Step 2: Look at some samples

In [52]:
print("\nSample of rows with missing Customer Names:")
print(df[df['Customer Name'].isna()][['Customer Name', 'State', 'Product', 'Total Sale', 'Order ID']].head())


Sample of rows with missing Customer Names:
   Customer Name    State     Product Sales Channel  Total Sale  \
0            NaN   Rivers    Keyboard        Online         NaN   
13           NaN     Kano      Camera           NaN         NaN   
26           NaN  Anambra     Monitor        Retail         NaN   
39           NaN     Osun      Tablet        Online         NaN   
52           NaN     Osun  Headphones     Wholesale         NaN   

                                Order ID  
0                                    NaN  
13  6d7df5d0-d3b5-4af7-a7ce-378fe5a4983b  
26  b0b6dedf-4f9c-4027-8514-4d98b41c504f  
39  d42a895e-0fcf-401f-9cb9-b73a906d5c8c  
52  6c776e95-0b98-4ccd-92b2-102d940c3503  


## Step 3: Think about it

## Step 4: Investigate the Correlation

In [57]:
# Check: Do rows with missing Customer Name also have missing Total Sale?
missing_customer = df['Customer Name'].isna()

print("Rows with missing Customer Name:")
print(f"Total: {missing_customer.sum()}")
print(f"\nOf those, how many ALSO have missing Total Sale?")
print(f"Count: {df[missing_customer]['Total Sale'].isna().sum()}")
print(f"Percentage: {df[missing_customer]['Total Sale'].isna().sum()/missing_customer.sum() * 100 :.2f}")

print("\n" + "="*50)
print("Let's look at the the missing customer rows in detail:")
print("="*50)
missing_customer_df = df[missing_customer][['Customer Name', 'State', 'Product', 'Units Sold', 'Unit Price', 'Total Sale', 'Sales Channel', 'Order ID']]
print(missing_customer_df)

print("\n" + "="*50)
print("Summary of missing values in rows with no Customer Name:")
print("="*50)
print(missing_customer_df.isna().sum())

Rows with missing Customer Name:
Total: 43

Of those, how many ALSO have missing Total Sale?
Count: 36
Percentage: 83.72

Let's look at the the missing customer rows in detail:
    Customer Name        State     Product  Units Sold  Unit Price  \
0             NaN       Rivers    Keyboard         NaN         NaN   
13            NaN         Kano      Camera         NaN   169095.82   
26            NaN      Anambra     Monitor         NaN   196849.50   
39            NaN         Osun      Tablet         NaN   171984.80   
52            NaN         Osun  Headphones         NaN   218356.93   
65            NaN         Osun     Monitor         NaN   260362.49   
78            NaN        Benue       Phone         NaN   207226.85   
91            NaN      Plateau      Laptop         NaN   103867.84   
104           NaN      Plateau     Monitor         1.0   168824.40   
117           NaN       Kaduna       Phone        77.0   204517.00   
130           NaN          Imo     Monitor         Na

## Step 5: Final Decision - Drop incomplete records

In [62]:
# Drop rows where both Customer Name and Total Sale is missing

# Create filter for rows to drop
incomplete_records = df['Customer Name'].isna() & df['Total Sale'].isna()

print(f"Records to drop: {incomplete_records.sum()}")
print(f"Records to keep: {(~incomplete_records).sum()}")

# Drop them
df_clean = df[~incomplete_records].copy()

# Verify
print(f"\nOriginal dataset: {len(df)} rows")
print(f"Cleaned dataset: {len(df_clean)} rows")
print(f"Rows removed: {len(df) - len(df_clean)} rows")
print(f"Percentage removed: {(len(df)-len(df_clean))/len(df) * 100:.2f}%")

Records to drop: 36
Records to keep: 514

Original dataset: 550 rows
Cleaned dataset: 514 rows
Rows removed: 36 rows
Percentage removed: 6.55%


# DAY 2 STRING CLEANING - FINAL SUMMARY
## Nigerian Sales Dataset

---

## CLEANING SUMMARY

### 1. STATE STANDARDIZATION
- **Before:** 40 unique values (case inconsistency)
- **After:** 20 unique values (Title Case)
- **Improvement:** 50% reduction in duplicates

### 2. PRODUCT STANDARDIZATION
- **Before:** 16 unique values (case inconsistency)
- **After:** 8 unique values (Title Case)
- **Improvement:** 50% reduction in duplicates

### 3. SALES CHANNEL
- **Status:** Already standardized (Title Case)
- **Missing:** 106 values (19.3%) - kept as NaN
- **Decision:** Leave missing for now, investigate source

### 4. INCOMPLETE RECORDS REMOVAL
- **Records analyzed:** 43 missing Customer Names
- **Correlation found:** 83% also missing Total Sale
- **Action:** Removed 36 incomplete records (6.55%)
- **Rationale:** Sales without amounts not useful for analysis

---

## FINAL RESULTS

**FINAL DATASET:** 514 rows (from 550 original)

---

## DATA QUALITY IMPROVEMENTS

✅ Standardized geographic data (States)  
✅ Standardized product categories  
✅ Removed incomplete transactions  
✅ Preserved data integrity (93.45% retention)

---

## KEY LEARNINGS

- Always inspect before cleaning  
- Case standardization eliminates artificial duplicates  
- Missing data requires investigation, not assumption  
- Correlation analysis reveals data quality patterns  
- Business context drives cleaning decisions

---

## SESSION STATS

- **Time invested:** ~45 minutes
- **Columns cleaned:** 4 (State, Product, Sales Channel, Customer Name)
- **Data quality improvement:** Significant
- **Skills practiced:** String methods, missing value analysis, decision-making